src/handlers.rs

```diff
@@ -262,8 +384,194 @@ fn serialize_initgroups(groups: Vec<Gid>) -> Result<Vec<u8>> {
     Ok(result)
 }
 
+impl Hostent {
+    fn serialize(&self) -> Result<Vec<u8>> {
+        // Loop over all addresses.
+        // Serialize them into a slice, which is used later in the payload.
+        // Take note of the number of addresses (by AF).
+        let mut num_v4 = 0;
+        let mut num_v6 = 0;
+        let mut buf_addrs = vec![];
+        let mut buf_aliases = vec![];
+        // Memory segment used to convey the size of the different
+        // aliases. The sizes are expressed in native endian encoded 32
+        // bits integer.
+        let mut buf_aliases_size = vec![];
+
+        for address in self.addr_list.iter() {
+            match address {
+                IpAddr::V4(ip4) => {
+                    num_v4 += 1;
+                    for octet in ip4.octets() {
+                        buf_addrs.push(octet)
+                    }
+                }
+                IpAddr::V6(ip6) => {
+                    num_v6 += 1;
+                    for octet in ip6.octets() {
+                        buf_addrs.push(octet)
+                    }
+                }
+            }
+        }
+
+        for alias in self.aliases.iter() {
+            let alias_bytes = CString::new(alias.clone())?.into_bytes_with_nul();
+            let size_in_bytes: i32 = alias_bytes.len() as i32;
+            buf_aliases_size.extend_from_slice(&size_in_bytes.to_ne_bytes());
+            buf_aliases.extend_from_slice(alias_bytes.as_slice());
+        }
+
+        // this can only ever express one address family
+        if num_v4 != 0 && num_v6 != 0 {
+            bail!("unable to serialize mixed AF")
+        }
+
+        let num_addrs = num_v4 + num_v6;
+        let has_addrs = num_addrs > 0;
+
+        let hostname_c_string_bytes = CString::new(self.name.clone())?.into_bytes_with_nul();
+        let hostname_c_string_len = if has_addrs {
+            hostname_c_string_bytes.len() as i32
+        } else {
+            0
+        };
+
+        let header = protocol::HstResponseHeader {
```

Comment:
I think this could be simplified by adding a `HST_RESPONSE_HEADER_NOT_FOUND` like you did for `AI_RESPONSE_HEADER_NOT_FOUND` - get rid of all the `has_addrs` checks. Unless `aliases` can still be valid or something?